### PR TITLE
Better download_and_gen script, support for perfmon-metrics

### PR DESCRIPTION
**Pull request by captain5050**
Hi,

these changes refactor existing converter scripts to be usable as python modules. A new converter script is added written in python. The new script automates the creation of the perf mapfile.csv highlighting some errors in upstream Linux. The new script downloads files using the descriptions in mapfile.csv, this means version changes don't need to be made to the script. The script uses Python's urllib library to avoid downloading then processing files, processing is done on the in-memory version. Perfmon-metrics events are added except in a few cases where errors were detected with 'perf test' in the github version and because of the known #SYSTEM_TSC_FREQ issue. Once errors in perfmon-metrics are fixed it is straightforward to enable generation of the metrics again.

The output of the new script matches the previous download_and_gen.sh except, the addition of mapfile.csv and perfmon-metrics, and uncore events are generated for architectures where no csv file is present  (by passing an empty csv file). As such the output is the same or superior.

Thanks,
Ian

### CODE REVIEW

File: extract-tma-metrics.py [hunk 1]
```diff
@@ -33,500 +33,604 @@
...
+            'IVB', 'SNB'),
+    'TGL': ('TGL', 'RKL', 'ICL', 'CNL', 'KBLR/CFL/CML', 'SKL/KBL', 'BDW/BDW-DE',
+            'HSW', 'IVB', 'SNB'),
+    'ADL/RPL': ('ADL/RPL', 'TGL', 'RKL', 'ICL', 'CNL', 'KBLR/CFL/CML',
+                'SKL/KBL', 'BDW', 'HSW', 'IVB', 'SNB'),
+    'SPR': ('SPR', 'ADL/RPL', 'TGL', 'RKL', 'ICX', 'ICL', 'CNL', 'CPX', 'CLX',
+            'KBLR/CFL/CML', 'SKX', 'SKL/KBL', 'BDX', 'BDW', 'HSX', 'HSW', 'IVT',
+            'IVB', 'JKT/SNB-EP', 'SNB'),
+    'GRT': ('GRT'),
 }
```
Comment (ZhengjunXing0) [inline extract-tma-metrics.py:136]: The old one looks better, why change this?

File: extract-tma-metrics.py [hunk 2]
```diff
@@ -33,500 +33,604 @@
...
+    ], [3, 6, 7], [2, 3, 6, 7]),
+    (['KBL'], [3, 6, 7], [2, 3, 6, 7]),
+    (['CNL'], [1, 3, 6, 7], [2, 3, 6, 7, 8, 9, 10]),
+    (['ICL', 'TGL', 'RKL'], [6, 7], [2, 3, 6, 7, 8, 9, 10]),
+    (['ICX', 'SPR'], [1, 6], [2, 6]),
+    (['ADL', 'GRT'], [1, 6, 7], [2, 3, 6, 7, 8, 9, 10]),
+    (['SLM'], [1, 6], [6]),
+    (['KNL', 'KNM'], [6], [2, 3, 6]),
+    (['GLM', 'SNR'], [1, 3, 6], [2, 3, 6, 10]),
 ]
```
Comment (ZhengjunXing0) [inline extract-tma-metrics.py:151]: The old one looks better, why change this?

File: extract-tma-metrics.py (hunk 1 above)
Comment (captain5050) [reply extract-tma-metrics.py:136]: Generally at Google we use automatic code formatting, it avoids arguments over white space and lets you focus on the substance of the change. The change here was caused by pyformat.

File: extract-tma-metrics.py (hunk 2 above)
Comment (captain5050) [reply extract-tma-metrics.py:151]: pyformat. We can hand indent but there were issues with line length, etc. with the old code.

**Comment by ZhengjunXing0** [comment]
Applied, thanks.

**Pull request merged**